### ISSUE TITLE
Generate invariants for C programs with the functions from math.h

### ISSUE THREAD

**Issue by TaihuLight** [unsat/dig#22]
`sqrt()' is a function defined in 'math.h' for C programs. Unfortunately, when I generate the invariants for the following C program, it occurs the error: undefined reference to `sqrt'
How to solve this problem? @nguyenthanhvuh 
`
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

void vassume(int b){}
void vtrace1(int n, int dd, int r, int rp, int q, int d, int s, int t){} 
void vtrace2(int n, int dd, int r, int rp, int q, int d, int s, int t){} 
void vtrace3(int n, int dd, int r, int rp, int q, int d, int s, int t){} 
void vtrace4(int n, int dd, int r, int rp, int q, int d, int s, int t){} 

void mainQ(int n, int dd) {
	int r, rp, q, d, s, t;
	d = dd;
	if (d!=0 && d!=2){
	r = n % d;
	t=r;
	rp = n % (d - 2);
	q = 4*(n / (d-2) - n / d);
	s = (int) sqrt(n);
	}

	while (s >= d && r != 0) {// loop invariant:
		if(2*r-rp+q<0){
		vtrace1(n, dd, r, rp, q, d, s, t);
			t = r;
			r = 2*r-rp+q+d+2;
			rp = t;
			q = q+4;
			d = d+2;
		} else if((2*r-rp+q >=0)&&(2*r-rp+q<d+2)){
		vtrace2(n, dd, r, rp, q, d, s, t);
			t = r;
			r = 2*r-rp+q;
			rp = t;
			d = d+2;
		} else if((2*r-rp+q >=0)&&(2*r-rp+q>=d+2)
			&&(2*r-rp+q <2*d+4)) {
			vtrace3(n, dd, r, rp, q, d, s, t);
			t = r;
			r = 2*r-rp+q-d -2;
			rp = t;
			q = q -4;
			d = d+2;
		} else{
			// ((2*r-rp+q>=0)&&(2*r-rp+q>=2*d+4)) 
			vtrace4(n, dd, r, rp, q, d, s, t);
			t = r;
			r = 2*r-rp+q -2*d -4;
			rp = t;
			q = q -8;
			d = d+2;
		}
	}
}

void main(int argc, char **argv){
    mainQ(atoi(argv[1]),atoi(argv[2]));
}
`

**Comment by nguyenthanhvuh** [comment]
Yes, unfortunately while pure dynamic invariant generation doesn't have problem with library function like sqrt,  DIG relies on symbolic execution and such to check the results and they might have problem with these functions.  A work around is using a nonsqrt version:  https://github.com/unsat/dig/blob/dev/benchmark/c/nla_full/knuth_nosqrt.c

**Comment by TaihuLight** [comment]
The newly updated DIG can remove redundancy from `eqts`. **However, it does not support removing redundancy from `ineqts`.** For the program, e.g, 

```
#include <stdio.h>
#include <stdlib.h>

void vassume(int b){}
void vtrace(int x, int y){} 

void mainQ( ) 
{
  int x,y;  
  x=0;
  y=0;
  vassume (x==0&&y==0);  
  while(x<=100)
    {// loop invariant:x==y
      vtrace(x,y);
      y=y+1;
      x=x+1;
    } 
    //assert(x==y);
}

void main(int argc, char **argv){
    mainQ();
}
```
the invariants generated with DIG is 
```
sage -python -O dig.py bench16_potsumm1.c -log 3 -nominmax -nominmaxplus -nomp -maxdeg 5 -seed 2
settings:INFO:2021-07-02 15:25:23.002700: dig.py bench16_potsumm1.c -log 3 -nominmax -nominmaxplus -nomp -maxdeg 5 -seed 2
alg:INFO:analyze '/home/ly/Music/benchmark-tasep/bench16_potsumm1.c'
alg:INFO:compute symbolic states (4.63s)
alg:INFO:infer invs at 1 locs: vtrace
alg:INFO:found 1 eqts (1.41s)
alg:INFO:found 5 ieqs (0.12s)
alg:INFO:test 6 invs using 101 traces (0.02s)
alg:INFO:simplify 6 invs (0.04s)
* prog bench16_potsumm1 locs 1;  invs 4 (Eqt: 1, Oct: 3) V 2 T 2 D 1;  NL 0 (1) ;
-> time eqts 1.4s, ieqs 0.1s, simplify 0.1s, symbolic_states 4.6s, total 6.2s
-> checks  0 () change depths 0 () change vals 0 ()
-> max  0 () change depths 0 ()
runs 1
rand seed 2.0, test (11, 57)
vtrace (4 invs):
1. x - y == 0
2. -y <= 0
3. x - y <= 0
4. -x + y <= 0
tmpdir: /var/tmp/dig_2_lke27k9i
```
The 3th and 4th invariants are equal to the 1st one, then they are redundant and can be removed.
(1) Does DIG support removing redundancy from `ineqts` (e.g., Removing x-y<=0 and -x+y<=0 for the above result)? 
(2) If not, will DIG support this operation in the coming days?

**Comment by nguyenthanhvuh** [comment]
This is by design.  In many cases the redundant invariants are still useful and not easily seen from the stronger, independent ones.  So if I just eliminate all redundant ones  the developers sometimes don't see what they expect.  Given that the total number of generated invariants is not too long,  I've decided to keep the equalities and some shorter inequalities, even though they might be redundant.
